### ISSUE TITLE
[HOLD for payment 2022-09-08] $500 - Audit forms and fix inconsistencies with focus, tab and shift + tab behavior 4/4

### ISSUE THREAD

**Issue by dylanexpensify** [Expensify/App#7918]
We should audit all our forms and fix any inconsistencies with focus, tab, shift + tab and enter behavior. The expected behavior is as follows:

1. `Tab` navigates to the next input.
2. `Shift + tab` navigates to the previous input.
3. `Enter` submits the form.
4. `Space` toggles checkboxes/dropdowns.

**Note:** We should make sure that tabbing cycles through the form in an order that makes sense, usually top to bottom.

Here's a list of forms to be audited:
- [ ] [Add Personal Bank account flow](https://github.com/Expensify/App/blob/30118b75ba9693ee5518464c8d7073a9c6883b19/src/components/AddPlaidBankAccount.js)
- [ ] [Settings - Change Password](https://github.com/Expensify/App/blob/30118b75ba9693ee5518464c8d7073a9c6883b19/src/pages/settings/PasswordPage.js)
- [ ] [Settings - Profile](https://github.com/Expensify/App/blob/30118b75ba9693ee5518464c8d7073a9c6883b19/src/pages/settings/Profile/ProfilePage.js)
- [ ] [Login form](https://github.com/Expensify/App/blob/main/src/pages/signin/LoginForm.js)
- [ ] [Request a Call modal](https://github.com/Expensify/App/blob/main/src/pages/RequestCallPage.js)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @MitchExpensify (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @MonilBhavsar (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by dylanexpensify** [comment]
Note: decision to split this main issue up into smaller issues came from [this convo](https://expensify.slack.com/archives/C01SKUP7QR0/p1645014281039819)

**Comment by MitchExpensify** [comment]
How does that change the normal contributor management process @dylanexpensify ? Thanks

**Comment by MelvinBot** [comment]
@MonilBhavsar, @MitchExpensify Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by MitchExpensify** [comment]
Bumping the above @dylanexpensify

**Comment by dylanexpensify** [comment]
Ah, we basically decided to break the issues into four smaller issues since we weren't sure about milestones. So now just proceed as normal, posting onto Upwork, adding exported, etc!

**Comment by MonilBhavsar** [comment]
Is this a part of "New Expensify forms"?

**Comment by MitchExpensify** [comment]
[Upwork Job](https://www.upwork.com/jobs/~0179faa3fd71c4c21b)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @pecanoro (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by mdneyazahmad** [comment]
## Proposal

I will wrap each modal page in `FocusTrap` component such that when modal opens. It does not focus elements outside of modal. And store the reference of last element clicked to open the modal so that when modal closes it focus on that element.  And of course audit and fix inconsistency.

I did it only for the `ProfilePage` Here is the implementation of `FocusTrap` Component Wrapper.
```javascript
import React, {useRef, useEffect} from 'react';
import {View} from 'react-native';
import PropTypes from 'prop-types';
import FocusTrapLib from 'focus-trap-react';

import { useIsFocused } from '@react-navigation/native';

const propTypes = {
    children: PropTypes.node.isRequired,
};


function FocusTrap(props) {
    const isFocused = useIsFocused();
    const lastFocusedElement = useRef(null);

    useEffect(() => {
        if(isFocused) {
            lastFocusedElement.current = document.activeElement;
            return;
        }

        if(lastFocusedElement.current) {
            lastFocusedElement.current.focus();
        }

    }, [isFocused]);

    return (
        <FocusTrapLib active={isFocused}>
            <View>
                {props.children}
            </View>
        </FocusTrapLib>
    );
}

FocusTrap.propTypes = propTypes;
FocusTrap.displayName = 'FocusTrap';

export default FocusTrap;

```

## Result

https://user-images.githubusercontent.com/77761491/157521857-2fbecc95-eea5-44c1-9b4a-0fc9cb40162a.mov

**Comment by mdneyazahmad** [comment]
This is not the final code there will be some refactoring. This is the base minimum to propose a solution. And also every other components will be refactored the same way.

**Comment by mdneyazahmad** [comment]
And also need to add `aria-role="modal'` and other aria related properties to the root of the modal component.

**Comment by parasharrajat** [comment]
What does FocusTrap do? Can we implement the minimal required functionality in our app?

**Comment by parasharrajat** [comment]
@luacmartins Could you please suggest if we are fine with focus-trapping in our pages? I think It would be needed to move the focus to RHN.

**Comment by mdneyazahmad** [comment]
@parasharrajat I also think the same we should trap it and it is the recommended pattern for a modal. Consider the case when you open a modal and press multiple tabs. The focus will be lost and will be focused on some another element that UI does not allow to interact with.


https://user-images.githubusercontent.com/77761491/157665747-fe51984a-7b87-45ac-a5a9-012bbb9364e4.mov

**Comment by parasharrajat** [comment]
Great. I got a yes for focus trapping. 

@mdneyazahmad 
> What does FocusTrap do? Can we implement the minimal required functionality in our app with using that lib?

**Comment by luacmartins** [comment]
>@luacmartins Could you please suggest if we are fine with focus-trapping in our pages? I think It would be needed to move the focus to RHN.

I think it makes sense to use focus trapping with the RHN since it blocks the rest of the app until we dismiss it.

**Comment by mdneyazahmad** [comment]
Here's the demo https://www.w3.org/TR/wai-aria-practices/examples/dialog-modal/dialog.html

**Comment by mdneyazahmad** [comment]
Yes, we can implement this feature it is not that big but the library covers more edge cases.

**Comment by parasharrajat** [comment]
I saw that lib has many extra features that we don't really need. It is not much in size but still.. 
![image](https://user-images.githubusercontent.com/24370807/157720048-013c031e-119c-402f-b35d-1d366bcd2828.png)


What do you suggest? Could you please tell me the reasons why we should use the lib? If you are fine implementing, can you please share the pusodo code here?

**Comment by mdneyazahmad** [comment]
@luacmartins does it mean we should use this library?

**Comment by parasharrajat** [comment]
> does it mean we should use this library?
Exactly my question as well. Haha...

**Comment by luacmartins** [comment]
That means we should avoid using an external library for this. Sorry if that wasn't clear 😅

**Comment by mdneyazahmad** [comment]
#### Algorithms:

1. When a modals opens, we will find the first focusable element and focus it.
2. `tab` follows it until we hit the last element. We check if the focused element is outside the dom node.

```
Case1 : not outside
        let it focus the next element.
Case2:  outside 
      There will be two case
       Case 1: focus is on the last element (tab pressed)
             focus the first element of modal
       Case 2: focus is on the first element (shift + tab pressed)
              focus the last element of modal
```

Please, give me some time I will share the implementation.

cc: @parasharrajat

**Comment by parasharrajat** [comment]
No worries. Take your time. I will make sure that you are assigned to another issue as well if you submit a good proposal.  

> When a modal opens, we will find the first focusable element and focus it.

I don't think we need this. Just the focus should cycle as you said. And on the first tab press, the focus should be on the first focusable element on the page and follow the proper focus pattern of the browser.

**Comment by mdneyazahmad** [comment]
This guidelines explains the modal behaviour https://www.w3.org/TR/wai-aria-practices-1.1/#dialog_modal
Do you think we should implement as described?

cc: @parasharrajat

**Comment by parasharrajat** [comment]
I don't have a preference. I just don't see that as part of the requirements but this is a good question to @luacmartins :bow:

**Comment by luacmartins** [comment]
@mdneyazahmad sorry, what exactly are you suggesting that we implement? Focusing on the first focusable element?

**Comment by MitchExpensify** [comment]
Price doubled to $500 FYI

**Comment by parasharrajat** [comment]
@luacmartins Should we be worried about focus trapping in this issue? I saw that other issues are only doing audits.

**Comment by luacmartins** [comment]
Let's not worry about focus trapping in this issue.

**Comment by mdneyazahmad** [comment]
## Proposal

#### Add Personal Bank account flow: 
need help navigating the page

#### Settings - Change Password:
1. `tab` and `shift` + `tab` work as expected.
2. `enter` submits form.
3. `eye` icon is toggled with both `enter` and `space`.
4. When we submit form and it has error it displays the error but does not focus on the error text input.

#### Settings - Profile:
1. `tab` and `shift` + `tab` work as expected.
2. `enter` submits form.
3. Checkbox and it's label is focussed separately.
4. Checkbox is toggled with `enter` and when pressed, it toggles and submits the form.
5. When `enter` is pressed on `add phone number` it opens the dialog as well as submits the form.

#### Login form:
1. `tab` and `shift` + `tab` work as expected.
2. `enter` submits form.
3. `eye` icon is toggled with both `enter` and `space`.
4. When a wrong password is enter and server produces error it is displayed and the text input is not outlined in red color.

#### Request a Call modal:
1. `tab` and `shift` + `tab` work as expected.
2. `enter` does not submit form.
3. Error message is shown on blur instead of on submit.
5. When we submit form and it has error it displays the error but does not focus on the error text input.

#### Suggestions
1. Checkbox should be toggled with only `space`.
2. Checkbox label should not focusable.
3. When an error appears in a form (on submit) it should focus on the first error input element.

cc: @parasharrajat

**Comment by parasharrajat** [comment]
I am fine with @mdneyazahmad doing the audit and fixes for this job. We can go through the suggestion on the PR.

cc: @pecanoro 

:ribbon: :eyes: :ribbon:  C+ reviewed

**Comment by MelvinBot** [comment]
📣 @mdneyazahmad You have been assigned to this job by @pecanoro!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md) 📖

**Comment by pecanoro** [comment]
@mdneyazahmad Feel free to apply for the job in Upwork and start working on a PR for this!

**Comment by MitchExpensify** [comment]
Hired @parasharrajat for eventual C+ payment
@mdneyazahmad ready to accept your proposal once made! Thanks

**Comment by mdneyazahmad** [comment]
Update: I will create pr by tomorrow.

**Comment by mdneyazahmad** [comment]
I am not sure what to create in pr `tab` and `tab + shift` and `enter` works as expected. The last time I created the audit `Request a Call modal` was not submitting on `enter` press but now it is fixed here in this commit https://github.com/Expensify/App/commit/c2218bc1de64ad83726b0dfdc41559ecedec89c0

I still need help in navigating the first page [Add Personal Bank account flow](https://github.com/Expensify/App/blob/30118b75ba9693ee5518464c8d7073a9c6883b19/src/components/AddPlaidBankAccount.js). Could you tell me the steps to navigate to this page or even do I have access to it?

cc: @parasharrajat

**Comment by parasharrajat** [comment]
You can go to Settings > Payments page. Then Click Add payment method > Add bank account. Then follow on till the end when the bank account is added to the account. Use testing creads for Plaid which you can get from their docs.

**Comment by mdneyazahmad** [comment]
It does not open (dev env) the flow for me results `cors` error.

<img width="470" alt="Screenshot 2022-04-01 at 8 16 23 AM" src="https://user-images.githubusercontent.com/77761491/161185371-79df9f77-1131-4645-b4d9-48819e31e79b.png">

**Comment by parasharrajat** [comment]
I think you will have to use staging. I don't know how to solve the cors issue. I also face this.

**Comment by luacmartins** [comment]
Hmm maybe [this thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1646167216688449?thread_ts=1646163503.484199&cid=C01GTK53T8Q) might help?

**Comment by pecanoro** [comment]
Not overdue, PR in progress

**Comment by mdneyazahmad** [comment]
### Issue

We have an issue with the `Button` component. This registers a global event listener for `enter` key press, when a button with `pressOnEnter` is mounted. It submits the form even when we are not pressing the button.

https://user-images.githubusercontent.com/77761491/162951216-c57e89c1-f27c-47ee-8ff4-dd5fca36bf2f.mp4


https://github.com/Expensify/App/blob/fd0f83a2e557c9f6d1adcf5be6feb761cd9f40b4/src/components/Button.js#L118-L132

### Suggested solution

1. Add local event listener for `enter` keypress may be we create a `HotKeys` component and wrap the form element inside it. This is triggered only when an enter is pressed from its children component. Doing this way we still have some more issues as `TextInput (react-native-web)` stops event propagation and it never trigger `SubmitFunction`.
```
<HotKeys keys="enter" onPressed={SubmitFunction}>
      // form element goes here
</HotKeys>
```

2. Add one more prop like `onSubmitEditing` on `TextInput` to `Checkbox` and `Select` component. It will listen for `enter` press event internally and calls `onSubmitEditing`.

I personally like 2nd approach, if we go with the first option we can not listen for the event `HotKeys` when it is pressed on `TextInput`.

cc: @parasharrajat

**Comment by parasharrajat** [comment]
Thanks for pointing it out. I am looking into that issue on another GH https://github.com/Expensify/App/issues/7623?notification_referrer_id=NT_kwDOAXPed7MzMDcyOTc3Njg4OjI0MzcwODA3&notifications_after=Y3Vyc29yOjI1#issuecomment-1085901192.

**Comment by luacmartins** [comment]
@mdneyazahmad please note that `4. Space toggles checkboxes.` was added to the list of expected behavior.

**Comment by mdneyazahmad** [comment]
This issue depends on https://github.com/Expensify/App/issues/7623 I will have to wait until that issue is fixed.

**Comment by MitchExpensify** [comment]
Not overdue waiting on #7623

**Comment by MitchExpensify** [comment]
Not overdue PR drafted

**Comment by mdneyazahmad** [comment]
Apologies for the delay on this one. This issue is on my high priority list and I want to finish it asap. It looks like the related issue(https://github.com/Expensify/App/issues/7918#issuecomment-1096631126) is not fixed. This issue is very common and not just related to this form. We need to differentiate between enter pressed on form element and a general button. Should we create a separate issue to handle this case?

If we add pressOnEnter here, it does not work. It has two steps enter email -> password. When we enter the email and press enter it will trigger pressOnEnter on the password step button too. We can not conditionally render the password step as specified in the comment.

```javascript
// file: src/pages/signin/SignInPage.js

<SignInPageLayout
                    welcomeText={welcomeText}
                    shouldShowWelcomeText={showLoginForm || showPasswordForm || !shouldShowResendValidationLinkForm}
                >
                    {/* LoginForm and PasswordForm must use the isVisible prop. This keeps them mounted, but visually hidden
                    so that password managers can access the values. Conditionally rendering these components will break this feature. */}
                    <LoginForm isVisible={showLoginForm} blurOnSubmit={shouldShowResendValidationLinkForm} />
                    <PasswordForm isVisible={showPasswordForm} />
                    {shouldShowResendValidationLinkForm && <ResendValidationForm />}
                </SignInPageLayout>

// file: src/pages/signin/PasswordForm.js

<Button
   pressOnEnter
/>
```
<img width="408" alt="Screenshot 2022-05-23 at 11 28 11 AM" src="https://user-images.githubusercontent.com/77761491/169753220-de2ecb3e-eb22-455e-9bb7-0c4385a3ff0b.png">


cc: @parasharrajat

**Comment by parasharrajat** [comment]
Use `isVisible` prop on the `LoginForm` and `PasswordForm` to control `pressOnEnter`.

**Comment by mdneyazahmad** [comment]
Sorry for the confusion 
```
<Button pressOnEnter={props.isVisible} />
````

When `pressOnEnter` is `true`, it registers shortcut when component is mounted (`componentDidMount`). When later on this prop is updated. It does not reregister or unregister.

**Comment by mdneyazahmad** [comment]
> What do you suggest to solve this issue?

I think, this is a complex issue and needs some discussion. May be, create a separate issue and see what proposal we get. It affects many part of the app. In fact, wherever we have `pressOnEnter`.

I think we can remove this prop (pressOnEnter) from button component and try to find some other method.

**Comment by parasharrajat** [comment]
> When pressOnEnter is true, it registers shortcut when the component is mounted (componentDidMount). When later on this prop is updated. It does not reregister or unregister.

Can't this be changed?

> this is a complex issue 

Hmm. :thinking:

**Comment by mdneyazahmad** [comment]
Yes, we can update the Button component to reregister.

**Comment by parasharrajat** [comment]
@mdneyazahmad Could you please share an update on the done things and the remaining ones?

**Comment by mdneyazahmad** [comment]
Same here. as https://github.com/Expensify/App/issues/7916#issuecomment-1136147023

**Comment by pecanoro** [comment]
Let me see if I understood the problem correctly, we can't use onEnter because it creates too many secondary unexpected behaviors, right? I am worried that if we remove pressOnEnter and we try to find another solution for only these cases, we are just looking for a hacky solution instead of solving it for all cases. @mdneyazahmad Can you bring up this issue in #expensify-open-source to see what the team thinks?

**Comment by MitchExpensify** [comment]
Not overdue, conversation ongoing in this [Slack thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1654195847140899)

**Comment by mdneyazahmad** [comment]
Not receiving any new thoughts on [Slack thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1654195847140899). Can someone help me with the next step?

**Comment by MitchExpensify** [comment]
Slack thread has been bumped

**Comment by parasharrajat** [comment]
Posted a possible solution on the slack thread.

**Comment by roryabraham** [comment]
Posted another possible solution on the slack thread and in [this draft PR](https://github.com/Expensify/App/pull/9749)

**Comment by parasharrajat** [comment]
@mdneyazahmad Could you respond on the slack thread and also try the above-mentioned solution. Let's keep this issue moving.

**Comment by MitchExpensify** [comment]
Doubled! New Upwork link [here](https://www.upwork.com/jobs/~01dd510b06d81cf27f)

**Comment by MitchExpensify** [comment]
My mistake @parasharrajat, fixed. Thanks for the catch

**Comment by MitchExpensify** [comment]
Mind making a proposal on the new listing @mdneyazahmad ? The old one expired

**Comment by parasharrajat** [comment]
I guess if they was hired for the last listing then they should have an active contract even though the listing is expired. In short, no need to apply again.

**Comment by MitchExpensify** [comment]
I'd prefer to manage payment centrally in the new listing @parasharrajat

**Comment by parasharrajat** [comment]
So the old contract needs to be cancelled as a duplicate.

**Comment by mdneyazahmad** [comment]
Discussion is on going in this slack [thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1654195847140899).

**Comment by MitchExpensify** [comment]
Bumped [the thread ](https://expensify.slack.com/archives/C01GTK53T8Q/p1660145223173299?thread_ts=1654195847.140899&cid=C01GTK53T8Q)

**Comment by MitchExpensify** [comment]
Consensus in the thread is to move forward with:

"Continue using a single global keyboard event handler library, patching it with the solution from [this PR](https://github.com/Expensify/App/pull/9749) so that local event handlers can override the global ones"

Sound good to you @mdneyazahmad ?

**Comment by mdneyazahmad** [comment]
@MitchExpensify

> "Continue using a single global keyboard event handler library, patching it with the solution from https://github.com/Expensify/App/pull/9749 so that local event handlers can override the global ones"

We are moving ahead with this option, any further comment related to this option will be made on [this PR](https://github.com/Expensify/App/pull/9749). Thank you.

**Comment by mdneyazahmad** [comment]
@parasharrajat As we were discussing the solution and the consensus is to solve as per the solution provided in this https://github.com/Expensify/App/pull/9749. If we make the pr close successfully. I will create the pr to fix all these audit fixes as the solution to all of this is going to be same.

**Comment by mdneyazahmad** [comment]
Kind of like that you can suggest a better approach.

**Comment by MitchExpensify** [comment]
What is the latest on this?

**Comment by MitchExpensify** [comment]
Not overdue, just on the lookout for regressions before paying out

**Comment by MitchExpensify** [comment]
Yes I closed it by mistake and reopened it afterward @mdneyazahmad

**Comment by MitchExpensify** [comment]
Waiting on [this](https://github.com/Expensify/App/issues/7918#issuecomment-1235481391)

**Comment by pecanoro** [comment]
Still waiting on https://github.com/Expensify/App/pull/10965

**Comment by pecanoro** [comment]
Not overdue, waiting on reviews

**Comment by pecanoro** [comment]
Still waiting on https://github.com/Expensify/App/pull/10965

**Comment by pecanoro** [comment]
Still waiting on https://github.com/Expensify/App/pull/10965

**Comment by pecanoro** [comment]
@mdneyazahmad what the last update on this?

**Comment by mdneyazahmad** [comment]
@pecanoro waiting on https://github.com/Expensify/App/pull/12370. This will be merged soon, then we can work this issue. Thanks!

**Comment by JmillsExpensify** [comment]
Still working on the linked PR, though in any case, I don't think this is a focus for WAQ per [this Slack thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1672088203968479). Adding a hold as a result.

**Comment by pecanoro** [comment]
Still on HOLD

**Comment by JmillsExpensify** [comment]
Taking this off hold.

**Comment by JmillsExpensify** [comment]
Note: Now that this is off hold, I think we have a decision before us. Do we do anything at all? I actually vote no, we shouldn't. I still don't think keyboard navigation is a priority – including but not limited to tab, shift + tab, etc. – so better to close this issue and move on. Please let me know if I'm missing something!

**Comment by luacmartins** [comment]
I think that's reasonable to me. The pages in this issue have either already changed ([Settings - Profile](https://github.com/Expensify/App/blob/30118b75ba9693ee5518464c8d7073a9c6883b19/src/pages/settings/Profile/ProfilePage.js)) or will be changed significantly as part of passwordless ([Settings - Change Password](https://github.com/Expensify/App/blob/30118b75ba9693ee5518464c8d7073a9c6883b19/src/pages/settings/PasswordPage.js), [Login form](https://github.com/Expensify/App/blob/main/src/pages/signin/LoginForm.js)), which would leave us with:

- [Request a Call modal](https://github.com/Expensify/App/blob/main/src/pages/RequestCallPage.js) - seems to work fine and I didn't see any issues with focus, tab and shift + tab behavior
- [Add Personal Bank account flow](https://github.com/Expensify/App/blob/30118b75ba9693ee5518464c8d7073a9c6883b19/src/components/AddPlaidBankAccount.js) - most of the flow is handled by Plaid, we only have a dropdown in our App. I didn't see any issues here either.

With that in mind, I think we might be good to close this one out!

**Comment by MitchExpensify** [comment]
Agree with closing for all the reasons stated above!

**Comment by JmillsExpensify** [comment]
Thanks for hearing me out!

**Comment by mdneyazahmad** [comment]
I have been assigned to this issue a long time ago, and have been constantly working on it. As this is a part of a larger issue broken into 4. I am assigned 3 of these including this. There are some challenges  to fix the issues, and there have been some discussions in the past that took a long time. There are two PR raised that fixes this issue https://github.com/Expensify/App/pull/8874 and https://github.com/Expensify/App/pull/12370. I think contributor involved in this issue should be paid. Thank you!

**Comment by pecanoro** [comment]
@MitchExpensify Can you open an internal discussion about the above in Slack? 🙇‍♀️

**Comment by MitchExpensify** [comment]
Ok to summarize how this played out to make sure I'm following:

Audit form inconsistencies were broken into four issue:

1. [Audit forms and fix inconsistencies with focus, tab and shift + tab behavior 1/4](https://github.com/Expensify/App/issues/7523)
2. [Audit forms and fix inconsistencies with focus, tab and shift + tab behavior 2/4](https://github.com/Expensify/App/issues/7916)
3. [Audit forms and fix inconsistencies with focus, tab and shift + tab behavior 3/4](https://github.com/Expensify/App/issues/7917)
4. [Audit forms and fix inconsistencies with focus, tab and shift + tab behavior 4/4](https://github.com/Expensify/App/issues/7918)

@mdneyazahmad is assigned to 2, 3, & 4 and received/will be receiving compensation for 2 & 3. With regards to compensation for 4 (this issue): We ultimately decided to nothing but work was completed (e.g. https://github.com/Expensify/App/pull/8874 and https://github.com/Expensify/App/pull/12370) that fixed inconsistencies flagged in the issue. As such, and for the work throughout being assigned to this issue, we need to agree on a fair figure.

Does that summarize where we're at? Just want to be sure we're on the same page before posting in Slack to land on the fair figure

**Comment by mdneyazahmad** [comment]
@MitchExpensify first of all all of these issues are of similar nature, maybe there was not require to break into 4, but it was broken into 4 to keep it simple.

1. As these are different issue, I think the price should not be adjusted with others otherwise no need to break.
2. As you can see the amount paid for 2/4 is just 250, but @parasharrajat better knows the effort and time required to complete that issue. Still I did not ask for a raise, because I thought, I can adjust with others.
3. The task required to do in this issue is to audit and fix any inconsistency. I have done the audit in the past. As these issues are similar. It has has to wait for the common PR. I would have started the 1st PR from any issue. But I thought 2/4 was little easy, therefore PR primarily addresses that issue.
4. We can also keep in mind the time this issues take.

I think the original price for this issue is justified. Maybe @parasharrajat has some input.


Thank you

**Comment by MitchExpensify** [comment]
Thank you both! I appreciate that extra context a lot. I've proposed this compensation in Slack and will issue payment accordingly!

**Comment by MitchExpensify** [comment]
Paid @parasharrajat via bonus on the old Upwork Job contract.

I've invite you, @mdneyazahmad, to the new [Upwork posting here](https://www.upwork.com/jobs/~0158331ee0f3860c7f) for payment seeing as the old one closed. Lemme know when you've accepted! Thanks for your

**Comment by mdneyazahmad** [comment]
@MitchExpensify accepted!

**Comment by MitchExpensify** [comment]
Awesome! Paid and contract ended. Thanks again everyone!
